File: src/test/java/com/soapdogg/lispInterpreter/parser/RootParserTest.java
package com.soapdogg.lispInterpreter.parser;

import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.datamodels.ParserResult;
import com.soapdogg.lispInterpreter.datamodels.Token;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RootParserTest {

    private Queue<Token> tokens;

    private Node resultingNode;

    private RootParser rootParser;

    @BeforeEach
    void setup() throws Exception {
        NodeParser nodeParser = Mockito.mock(NodeParser.class);

        Token headToken = Mockito.mock(Token.class);
        tokens = new LinkedList<>();
        tokens.add(headToken);

        resultingNode = Mockito.mock(Node.class);
        Queue<Token> remainingTokens = new LinkedList<>();

        ParserResult parserResult = Mockito.mock(ParserResult.class);
        Mockito.when(parserResult.getResultingNode()).thenReturn(resultingNode);
        Mockito.when(parserResult.getRemainingTokens()).thenReturn(remainingTokens);

        Mockito.when(nodeParser.parseIntoNode(tokens)).thenReturn(parserResult);

        rootParser = RootParser.newInstance(nodeParser);
    }

    @Test
    void rootParserTest() throws Exception {
        List<Node> actual = rootParser.parse(tokens);

        Assertions.assertEquals(1, actual.size());
        Assertions.assertEquals(resultingNode, actual.get(0));
    }
}

File: src/test/java/com/soapdogg/lispInterpreter/parser/RootParserTest.kt
package com.soapdogg.lispInterpreter.parser

import com.soapdogg.lispInterpreter.datamodels.Node
import com.soapdogg.lispInterpreter.datamodels.Token
import com.soapdogg.lispInterpreter.datamodels.TokenKind
import com.soapdogg.lispInterpreter.exceptions.UnexpectedTokenKindException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class RootParserTest {
    private val nodeParser = Mockito.mock(NodeParser::class.java)
    private val rootParser = RootParser(nodeParser)

    @Test
    fun rootParserTest() {
        val headToken = Mockito.mock(Token::class.java)
        val tokens = listOf(headToken)

        val resultingNode = Mockito.mock(Node::class.java)
        Mockito.`when`(nodeParser.parseIntoNode(tokens)).thenReturn(resultingNode)

        val actual = rootParser.parse(tokens)
        Assertions.assertEquals(1, actual.size)
        Assertions.assertEquals(resultingNode, actual[0])
    }

    @Test
    fun tooManyOpenTokenTest() {
        val headToken = Mockito.mock(Token::class.java)
        Mockito.`when`(headToken.tokenKind).thenReturn(TokenKind.OPEN_TOKEN)

        val tokens = listOf(headToken)

        Assertions.assertThrows(
            UnexpectedTokenKindException::class.java
        ) {rootParser.parse(tokens)}

        Mockito.verifyNoInteractions(nodeParser)
    }

    @Test
    fun tooManyCloseTokenTest() {
        val headToken = Mockito.mock(Token::class.java)
        Mockito.`when`(headToken.tokenKind).thenReturn(TokenKind.CLOSE_TOKEN)

        val tokens = listOf(headToken)

        Assertions.assertThrows(
            UnexpectedTokenKindException::class.java
        ) {rootParser.parse(tokens)}

        Mockito.verifyNoInteractions(nodeParser)
    }
}